Provide better messages when waiting for a condition in test

cpettitt-confluent opened the pull request:
This change lifts `retryOnExceptionWithTimeout` from `OptimizedKTableIntegrationTest` into `TestUtils` and uses it for some of the pre-canned condition wait functions. `retryOnExceptionWithTimeout` differs from `waitForCondition` in that it allows you to use regular assertions, with more descriptive messages and more details context when a test fails.

Some examples:

* `IntegrationTestUtils.waitUntilMinValuesRecordsReceived`

Before:

```
java.lang.AssertionError: Condition not met within timeout 1000. Did not receive all 51 records from topic output-three-0

	at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:377)
	at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:354)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilMinValuesRecordsReceived(IntegrationTestUtils.java:682)
	at org.apache.kafka.streams.integration.QueryableStateIntegrationTest.waitUntilAtLeastNumRecordProcessed(QueryableStateIntegrationTest.java:1000)
	at org.apache.kafka.streams.integration.QueryableStateIntegrationTest.queryOnRebalance(QueryableStateIntegrationTest.java:379)
```

After:

```
java.lang.AssertionError: Did not receive all 51 records from topic output-three-0 within 1000 ms
Expected: is a value equal to or greater than <51>
     but: <38> was less than <51>

	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.lambda$waitUntilMinValuesRecordsReceived$6(IntegrationTestUtils.java:674)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:417)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:385)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilMinValuesRecordsReceived(IntegrationTestUtils.java:670)
	at org.apache.kafka.streams.integration.QueryableStateIntegrationTest.waitUntilAtLeastNumRecordProcessed(QueryableStateIntegrationTest.java:1000)
	at org.apache.kafka.streams.integration.QueryableStateIntegrationTest.queryOnRebalance(QueryableStateIntegrationTest.java:379)
```

* `IntegrationTestUtils.waitUntilMinRecordsReceived`

Before:

```
java.lang.AssertionError: Condition not met within timeout 60000. Did not receive all 1 records from topic outputTopic

	at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:377)
	at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:354)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilMinRecordsReceived(IntegrationTestUtils.java:460)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilMinRecordsReceived(IntegrationTestUtils.java:432)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.verifyKeyValueTimestamps(IntegrationTestUtils.java:729)
	at org.apache.kafka.streams.integration.AbstractJoinIntegrationTest.checkResult(AbstractJoinIntegrationTest.java:170)
	at org.apache.kafka.streams.integration.AbstractJoinIntegrationTest.runTest(AbstractJoinIntegrationTest.java:218)
	at org.apache.kafka.streams.integration.TableTableJoinIntegrationTest.testInner(TableTableJoinIntegrationTest.java:121)
```

After:

```
java.lang.AssertionError: Did not receive all 51 records from topic outputTopic within 60000 ms
Expected: is a value equal to or greater than <51>
     but: <1> was less than <51>

	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.lambda$waitUntilMinRecordsReceived$0(IntegrationTestUtils.java:458)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:417)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:385)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilMinRecordsReceived(IntegrationTestUtils.java:454)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilMinRecordsReceived(IntegrationTestUtils.java:432)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.verifyKeyValueTimestamps(IntegrationTestUtils.java:734)
	at org.apache.kafka.streams.integration.AbstractJoinIntegrationTest.checkResult(AbstractJoinIntegrationTest.java:170)
	at org.apache.kafka.streams.integration.AbstractJoinIntegrationTest.runTest(AbstractJoinIntegrationTest.java:218)
	at org.apache.kafka.streams.integration.TableTableJoinIntegrationTest.testInner(TableTableJoinIntegrationTest.java:121)
```

* And my favorite: `IntegrationTestUtils.waitUntilFinalKeyValueRecordsReceived`

Before:

```
java.lang.AssertionError: Condition not met within timeout 60000. Did not receive all [KeyValueTimestamp{key=A, value=A:A, timestamp=1570731458595}, KeyValueTimestamp{key=B, value=B:B, timestamp=1570731458595}, KeyValueTimestamp{key=C, value=C:C, timestamp=1570731458595}, KeyValueTimestamp{key=D, value=D:D, timestamp=1570731458595}, KeyValueTimestamp{key=F, value=F:F, timestamp=1570731458595}, KeyValueTimestamp{key=E, value=E:E, timestamp=1570731458595}] records from topic output-0

	at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:377)
	at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:354)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilFinalKeyValueRecordsReceived(IntegrationTestUtils.java:638)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilFinalKeyValueTimestampRecordsReceived(IntegrationTestUtils.java:569)
	at org.apache.kafka.streams.integration.KStreamAggregationDedupIntegrationTest.validateReceivedMessages(KStreamAggregationDedupIntegrationTest.java:246)
	at org.apache.kafka.streams.integration.KStreamAggregationDedupIntegrationTest.shouldReduce(KStreamAggregationDedupIntegrationTest.java:125)
```

After:

```
java.lang.AssertionError: Received records from topic output-0 did not match expected within 60000 ms
Expected: is <{A=[KeyValueTimestamp{key=A, value=A:A, timestamp=1570731302494}], B=[KeyValueTimestamp{key=B, value=B:B, timestamp=1570731302494}], C=[KeyValueTimestamp{key=C, value=C:C, timestamp=1570731302494}], D=[KeyValueTimestamp{key=D, value=D:D, timestamp=1570731302494}], E=[KeyValueTimestamp{key=E, value=E:E, timestamp=1570731302494}], F=[KeyValueTimestamp{key=F, value=F:F, timestamp=1570731302494}]}>
     but: was <{A=[KeyValueTimestamp{key=A, value=A:A, timestamp=1570731302494}], B=[KeyValueTimestamp{key=B, value=B:B, timestamp=1570731302494}], C=[KeyValueTimestamp{key=C, value=C:C, timestamp=1570731302494}], D=[KeyValueTimestamp{key=D, value=D:D, timestamp=1570731302494}], E=[KeyValueTimestamp{key=E, value=E:E, timestamp=1570731302494}]}>

	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.lambda$waitUntilFinalKeyValueRecordsReceived$5(IntegrationTestUtils.java:631)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:417)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:385)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilFinalKeyValueRecordsReceived(IntegrationTestUtils.java:603)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilFinalKeyValueTimestampRecordsReceived(IntegrationTestUtils.java:566)
	at org.apache.kafka.streams.integration.KStreamAggregationDedupIntegrationTest.validateReceivedMessages(KStreamAggregationDedupIntegrationTest.java:246)
	at org.apache.kafka.streams.integration.KStreamAggregationDedupIntegrationTest.shouldReduce(KStreamAggregationDedupIntegrationTest.java:125)
```

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

abbccdda reviewed (commented):
Since we already work on this, does it make sense to also include a `diff` to the expected vs actual map?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:
```diff
@@ -16,6 +16,30 @@
  */
 package org.apache.kafka.streams.integration.utils;
 
+import static org.hamcrest.MatcherAssert.assertThat;
```
@cpettitt-confluent Can you change your IDE setting to void reformatting like this?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java:
```diff
@@ -189,7 +189,7 @@ public void shouldApplyUpdatesToStandbyStore() throws Exception {
         final ReadOnlyKeyValueStore<Integer, Integer> newActiveStore =
             kafkaStreams1WasFirstActive ? store2 : store1;
-        retryOnExceptionWithTimeout(100, 60 * 1000, TimeUnit.MILLISECONDS, () -> {
+        TestUtils.retryOnExceptionWithTimeout(100, 60 * 1000, () -> {
```
nit: add static import?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:
```diff
@@ -449,15 +449,14 @@ public static void waitForCompletion(final KafkaStreams streams,
+        final String reason = String.format("Did not receive all %d records from topic %s within %d ms", expectedNumRecords, topic, waitTime);
         try (final Consumer<K, V> consumer = createConsumer(consumerConfig)) {
-            final TestCondition valuesRead = () -> {
+            TestUtils.retryOnExceptionWithTimeout(waitTime, () -> {
```
nit: add static import? (similar elsewhere)

cpettitt-confluent commented:
It is probably safer not to slot this under `waitForCondition` because I've discovered we have some tests that actually use AssertionError to immediately abort, a la:

```
           TestUtils.waitForCondition(
                () -> {
                    try {
                        // ...
                    } catch (final InvalidStateStoreException e) {
                        // there must have been at least one rebalance state
                        assertTrue(stateListenerStub.mapStates.get(KafkaStreams.State.REBALANCING) >= 1);
                        return false;
                    }
                })
```

I'll keep the retry code for the specific cases where we know it is safe, but leave the old `waitForCondition` implementation for the more general cases where we may not want to retry on `AssertionError`.

I'll also follow up with fixes requested by @mjsax .

cpettitt-confluent committed:
Provide better messages when waiting for a condition in test

This change lifts `retryOnExceptionWithTimeout` from `OptimizedKTableIntegrationTest` into `TestUtils` and uses it for some of the pre-canned condition wait functions. `retryOnExceptionWithTimeout` differs from `waitForCondition` in that it allows you to use regular assertions, with more descriptive messages and more details context when a test fails.

Some examples:

* `IntegrationTestUtils.waitUntilMinValuesRecordsReceived`

Before:

```
java.lang.AssertionError: Condition not met within timeout 1000. Did not receive all 51 records from topic output-three-0

	at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:377)
	at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:354)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilMinValuesRecordsReceived(IntegrationTestUtils.java:682)
	at org.apache.kafka.streams.integration.QueryableStateIntegrationTest.waitUntilAtLeastNumRecordProcessed(QueryableStateIntegrationTest.java:1000)
	at org.apache.kafka.streams.integration.QueryableStateIntegrationTest.queryOnRebalance(QueryableStateIntegrationTest.java:379)
```

After:

```
java.lang.AssertionError: Did not receive all 51 records from topic output-three-0 within 1000 ms
Expected: is a value equal to or greater than <51>
     but: <38> was less than <51>

	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.lambda$waitUntilMinValuesRecordsReceived$6(IntegrationTestUtils.java:674)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:417)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:385)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilMinValuesRecordsReceived(IntegrationTestUtils.java:670)
	at org.apache.kafka.streams.integration.QueryableStateIntegrationTest.waitUntilAtLeastNumRecordProcessed(QueryableStateIntegrationTest.java:1000)
	at org.apache.kafka.streams.integration.QueryableStateIntegrationTest.queryOnRebalance(QueryableStateIntegrationTest.java:379)
```

* `IntegrationTestUtils.waitUntilMinRecordsReceived`

Before:

```
java.lang.AssertionError: Condition not met within timeout 60000. Did not receive all 1 records from topic outputTopic

	at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:377)
	at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:354)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilMinRecordsReceived(IntegrationTestUtils.java:460)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilMinRecordsReceived(IntegrationTestUtils.java:432)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.verifyKeyValueTimestamps(IntegrationTestUtils.java:729)
	at org.apache.kafka.streams.integration.AbstractJoinIntegrationTest.checkResult(AbstractJoinIntegrationTest.java:170)
	at org.apache.kafka.streams.integration.AbstractJoinIntegrationTest.runTest(AbstractJoinIntegrationTest.java:218)
	at org.apache.kafka.streams.integration.TableTableJoinIntegrationTest.testInner(TableTableJoinIntegrationTest.java:121)
```

After:

```
java.lang.AssertionError: Did not receive all 51 records from topic outputTopic within 60000 ms
Expected: is a value equal to or greater than <51>
     but: <1> was less than <51>

	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.lambda$waitUntilMinRecordsReceived$0(IntegrationTestUtils.java:458)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:417)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:385)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilMinRecordsReceived(IntegrationTestUtils.java:454)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilMinRecordsReceived(IntegrationTestUtils.java:432)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.verifyKeyValueTimestamps(IntegrationTestUtils.java:734)
	at org.apache.kafka.streams.integration.AbstractJoinIntegrationTest.checkResult(AbstractJoinIntegrationTest.java:170)
	at org.apache.kafka.streams.integration.AbstractJoinIntegrationTest.runTest(AbstractJoinIntegrationTest.java:218)
	at org.apache.kafka.streams.integration.TableTableJoinIntegrationTest.testInner(TableTableJoinIntegrationTest.java:121)
```

* And my favorite: `IntegrationTestUtils.waitUntilFinalKeyValueRecordsReceived`

Before:

```
java.lang.AssertionError: Condition not met within timeout 60000. Did not receive all [KeyValueTimestamp{key=A, value=A:A, timestamp=1570731458595}, KeyValueTimestamp{key=B, value=B:B, timestamp=1570731458595}, KeyValueTimestamp{key=C, value=C:C, timestamp=1570731458595}, KeyValueTimestamp{key=D, value=D:D, timestamp=1570731458595}, KeyValueTimestamp{key=F, value=F:F, timestamp=1570731458595}, KeyValueTimestamp{key=E, value=E:E, timestamp=1570731458595}] records from topic output-0

	at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:377)
	at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:354)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilFinalKeyValueRecordsReceived(IntegrationTestUtils.java:638)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilFinalKeyValueTimestampRecordsReceived(IntegrationTestUtils.java:569)
	at org.apache.kafka.streams.integration.KStreamAggregationDedupIntegrationTest.validateReceivedMessages(KStreamAggregationDedupIntegrationTest.java:246)
	at org.apache.kafka.streams.integration.KStreamAggregationDedupIntegrationTest.shouldReduce(KStreamAggregationDedupIntegrationTest.java:125)
```

After:

```
java.lang.AssertionError: Received records from topic output-0 did not match expected within 60000 ms
Expected: is <{A=[KeyValueTimestamp{key=A, value=A:A, timestamp=1570731302494}], B=[KeyValueTimestamp{key=B, value=B:B, timestamp=1570731302494}], C=[KeyValueTimestamp{key=C, value=C:C, timestamp=1570731302494}], D=[KeyValueTimestamp{key=D, value=D:D, timestamp=1570731302494}], E=[KeyValueTimestamp{key=E, value=E:E, timestamp=1570731302494}], F=[KeyValueTimestamp{key=F, value=F:F, timestamp=1570731302494}]}>
     but: was <{A=[KeyValueTimestamp{key=A, value=A:A, timestamp=1570731302494}], B=[KeyValueTimestamp{key=B, value=B:B, timestamp=1570731302494}], C=[KeyValueTimestamp{key=C, value=C:C, timestamp=1570731302494}], D=[KeyValueTimestamp{key=D, value=D:D, timestamp=1570731302494}], E=[KeyValueTimestamp{key=E, value=E:E, timestamp=1570731302494}]}>

	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.lambda$waitUntilFinalKeyValueRecordsReceived$5(IntegrationTestUtils.java:631)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:417)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:385)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilFinalKeyValueRecordsReceived(IntegrationTestUtils.java:603)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitUntilFinalKeyValueTimestampRecordsReceived(IntegrationTestUtils.java:566)
	at org.apache.kafka.streams.integration.KStreamAggregationDedupIntegrationTest.validateReceivedMessages(KStreamAggregationDedupIntegrationTest.java:246)
	at org.apache.kafka.streams.integration.KStreamAggregationDedupIntegrationTest.shouldReduce(KStreamAggregationDedupIntegrationTest.java:125)
```

cpettitt-confluent commented:
Changes up and ready for review.

abbccdda reviewed (commented):
Could you take a look at failed tests?

cpettitt-confluent commented:
Looks like hamcrest is not marked as a dependency in one of the modules. Reviews can be still be done in parallel. I'll fix the dependency later today.

bbejeck commented on clients/src/test/java/org/apache/kafka/test/TestUtils.java:
```diff
@@ -353,28 +355,77 @@ public static void waitForCondition(final TestCondition testCondition, final Sup
+                                                   final ValuelessCallable runnable) throws InterruptedException {
+        final long expectedEnd = System.currentTimeMillis() + timeoutMs;
+
+        while (true) {
```
IMHO it seems we may be susceptible to a hanging test here.  Should we change the condition of the `while` loop from to have a check that the `expectedEnd ` has elapsed?

cpettitt-confluent replied:
I don't want to put the condition in the while condition expression because I want to be sure the test runs at least once. I could add a separate boolean to check if we've run once, but don't want to overcomplicate it if not needed. I'm not sure why there is a risk of hanging?

We always hit runnable.call. Assuming it doesn't hang itself (which we can't do anything about, short of spinning up another thread), it will do one of:

1. Return successfully, in which case the whole function returns.
2. Throws an Exception, in which case we test if the timeout has expired and a) rethrow if it has or b) sleep and retest if it has not.
3. Throws an Assertion error, in which case we handled it the same way as 2.
4. Throws an Error, in which case the we also unwind the stack, just without rethrowing.

Let me know if I missed a case. Thanks!

cpettitt-confluent replied:
Also looking at it again, we would need to keep ahold of the last exception to be able to rethrow it when we exit the while loop - in fact that could act like our boolean. Still, I would prefer to keep this simple if possible. Please do let me know if I missed something in my analysis. Thanks again for the review!

bbejeck replied:
Thanks for the detailed explanation, SGTM

bbejeck reviewed (commented):
Hi @cpettitt-confluent thanks for the PR, I just one minor comment about the condition loop.

cpettitt-confluent committed:
Add hamcrest to streams-scala module

cpettitt-confluent commented:
Pushed a new commit that adds hamcrest to streams-scala. This should resolves the test failures we saw earlier.

cpettitt-confluent commented:
One test failure on 11/2.13 in kafka.log.LogCleanerIntegrationTest.testIsThreadFailed, which should be unrelated to this change. Other tests never kicked off? Retest this please.

cpettitt-confluent commented:
11/2.12 and 11/2.13 are clean.

8/2.11 has two test failures:

1. Known flaky test (https://issues.apache.org/jira/browse/KAFKA-8909): kafka.admin.ReassignPartitionsClusterTest.shouldPerformMultipleReassignmentOperationsOverVariousTopics.
2. kafka.log.LogCleanerIntegrationTest.testIsThreadFailed, which has not yet been reported as a flaky test.

retest this please.

cpettitt-confluent commented:
I'm going to dig into the LogCleanerIntegrationTest in parallel since it has failed twice and is not yet reported as a flaky test.

cpettitt-confluent commented:
My best guess is that the LogCleanerIntegrationTest failure is a flaky test in the making. It was apparently added about 3 days ago. Locally I've run the test on trunk and on my PR branch over 30 times with no failures.

mjsax reviewed (approved):
LGTM.

bbejeck commented:
Java 11/2.12 failed with `kafka.api.SaslSslAdminClientIntegrationTest.testCreateTopicsResponseMetadataAndConfig`

Java 11/2.13 failed with `kafka.admin.ReassignPartitionsClusterTest.shouldTriggerReassignmentWithZnodePrecedenceOnControllerStartup`

Java 8 passed.

retest this please

mjsax commented:
Java 8: `kafka.api.SslAdminClientIntegrationTest.testCreateTopicsResponseMetadataAndConfig`

Java 11 / 2.12:
```
kafka.log.LogCleanerIntegrationTest.testIsThreadFailed
org.apache.kafka.streams.state.internals.RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.shouldForwardAllDbOptionsCalls
```

Java 11 / 2.13:
`kafka.api.DelegationTokenEndToEndAuthorizationTest.testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl`

Retest this please.

mjsax commented:
Merged to `trunk` and cherry-picked to `2.4`.

Thanks for the cleanup/improvements @cpettitt-confluent!
